platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -0,0 +1,648 @@
+}
+
+void
+odp_crypto_compl_result(odp_crypto_compl_t completion_event,
+			odp_crypto_op_result_t *result)
+{
+	(void)completion_event;
+	(void)result;
+
+	/* We won't get such events anyway, so there can be no result */
+	ODP_ASSERT(0);
+}
+
+void
+odp_crypto_compl_free(odp_crypto_compl_t completion_event)
+{
+	odp_event_t ev = odp_crypto_compl_to_event(completion_event);
+
+	odp_buffer_free(odp_buffer_from_event(ev));
+}
+
+uint64_t odp_crypto_compl_to_u64(odp_crypto_compl_t hdl)
+{
+	return _odp_pri(hdl);
+}
+
+void odp_crypto_session_param_init(odp_crypto_session_param_t *param)
+{
+	memset(param, 0, sizeof(odp_crypto_session_param_t));
+}
+
+uint64_t odp_crypto_session_to_u64(odp_crypto_session_t hdl)
+{
+	return (uint64_t)hdl;
+}
+
+odp_packet_t odp_crypto_packet_from_event(odp_event_t ev)
+{
+	/* This check not mandated by the API specification */
+	ODP_ASSERT(odp_event_type(ev) == ODP_EVENT_PACKET);
+	ODP_ASSERT(odp_event_subtype(ev) == ODP_EVENT_PACKET_CRYPTO);
+
+	return odp_packet_from_event(ev);
+}
+
+odp_event_t odp_crypto_packet_to_event(odp_packet_t pkt)
+{
+	return odp_packet_to_event(pkt);
+}
+
+static
+odp_crypto_packet_result_t *get_op_result_from_packet(odp_packet_t pkt)
+{
+	odp_packet_hdr_t *hdr = packet_hdr(pkt);
+
+	return &hdr->crypto_op_result;
+}
+
+int odp_crypto_result(odp_crypto_packet_result_t *result,
+		      odp_packet_t packet)
+{
+	odp_crypto_packet_result_t *op_result;
+
+	ODP_ASSERT(odp_event_subtype(odp_packet_to_event(packet)) ==
+		   ODP_EVENT_PACKET_CRYPTO);
+
+	op_result = get_op_result_from_packet(packet);
+
+	memcpy(result, op_result, sizeof(*result));
+
+	return 0;
+}
+
+static
+int crypto_int(odp_packet_t pkt_in,
+	       odp_packet_t *pkt_out,
+	       const odp_crypto_packet_op_param_t *param)
+{
+	odp_crypto_alg_err_t rc_cipher = ODP_CRYPTO_ALG_ERR_NONE;
+	odp_crypto_alg_err_t rc_auth = ODP_CRYPTO_ALG_ERR_NONE;
+	odp_crypto_generic_session_t *session;
+	odp_bool_t allocated = false;
+	odp_packet_t out_pkt = *pkt_out;
+	odp_crypto_packet_result_t *op_result;
+	odp_packet_hdr_t *pkt_hdr;
+
+	session = (odp_crypto_generic_session_t *)(intptr_t)param->session;
+
+	/* Resolve output buffer */
+	if (ODP_PACKET_INVALID == out_pkt &&
+	    ODP_POOL_INVALID != session->p.output_pool) {
+		out_pkt = odp_packet_alloc(session->p.output_pool,
+					   odp_packet_len(pkt_in));
+		allocated = true;
+	}
+
+	if (odp_unlikely(ODP_PACKET_INVALID == out_pkt)) {
+		ODP_DBG("Alloc failed.\n");
+		return -1;
+	}
+
+	if (pkt_in != out_pkt) {
+		int ret;
+
+		ret = odp_packet_copy_from_pkt(out_pkt,
+					       0,
+					       pkt_in,
+					       0,
+					       odp_packet_len(pkt_in));
+		if (odp_unlikely(ret < 0))
+			goto err;
+
+		_odp_packet_copy_md_to_packet(pkt_in, out_pkt);
+		odp_packet_free(pkt_in);
+		pkt_in = ODP_PACKET_INVALID;
+	}
+
+	/* Order: Cipher followed by auth */
+	rc_cipher = session->cipher.func(out_pkt, param, session);
+	rc_auth = session->auth.func(out_pkt, param, session);
```

Comment:
This is not right. The order in decode must be the opposite of the order in encode, even if the implementation supported one ordering. And if the implementation supports one order only, then it should reject session creation when the other order is requested.

This does not matter much with AES-GCM but will break once individual cipher and auth algorithms get introduced.

Reply:
I am a bit confused as to how session creation can be rejected when the unsupported order is requested. Looking at the way this is handled in the `odp_crypto_openssl.c`, the `do_cipher_first` Boolean value in the `odp_crypto_generic_session_t` is assigned in session creation, depending on whether we are encoding or decoding:
```
        if (ODP_CRYPTO_OP_ENCODE == param->op)
                session->do_cipher_first = param->auth_cipher_text;
        else
                session->do_cipher_first = !param->auth_cipher_text;
```
This value is then checked in `crypto_int`:
```
/* Invoke the functions */
if (session->do_cipher_first) {
	rc_cipher = session->cipher.func(out_pkt, param, session);
	rc_auth = session->auth.func(out_pkt, param, session);
} else {
	rc_auth = session->auth.func(out_pkt, param, session);
	rc_cipher = session->cipher.func(out_pkt, param, session);
}
```
Can you explain more what change is required?

Reply:
> I am a bit confused as to how session creation can be rejected when the unsupported order is requested. 

By returning a negative value from odp_crypto_session_create() when a session cannot be created with the requested parameters.

odp_crypto_session_param_t contains auth_cipher_text parameter which specifies the order of ciphering and authentication when separate cipher and auth algorithms are used. If the session creation succeeds, the session has to work as specified in the parameters. You cannot apply the algorithms in a different order than requested as that would result in garbage output. It is just like you cannot just change the algorithm to something else if you did not like what what asked in the session creation parameters.

To comply with the ODP API, you should support both orders (the cipher text is authenticated or the plain text is authenticated). If you can support only one order, you should fail session creation if the other ordering is requested. But I do see why you could not support both orders.

Note that in your current code encode type sessions authenticate the cipher test but your decode type sessions authenticate the plain text. You probably did not intend that.

> Looking at the way this is handled in the odp_crypto_openssl.c

Yes, you can take a look at how it is handled there.

Reply:
This went wrong in two ways:

1) The code change belongs in the first patch. Now it is in the "add AES-GCM" patch even though it does not have anything to do with AES-GCM. And we do not want one patch to introduce incorrect code just to have a second patch to fix it. So make the change in the fist patch.

2) The logic is wrong. It seems you copied the code fragment from openssl crypto but for some reason reversed the logic of setting the do_cipher_first field. So now the order is wrong. The auth_cipher_text parameter means that the ciphered text (as opposed to plain text) is to be authenticated, i.e. ciphering must be done first in encode and last in decode, like the openssl crypto is doing.

Reply:
Yeah, I didn't realize that this change was merged in the second patch instead of the first patch. I will move it and change the order of the encode and decode.